File: sdk/typescript/src/webhook.ts
import { createHmac, timingSafeEqual } from 'node:crypto'
import { ImpriWebhookSignatureError } from './errors.js'

/**
 * Verify an Impri webhook delivery.
 *
 * Algorithm: HMAC-SHA256(secret, `${timestamp}.${nonce}.${rawBody}`)
 * compared constant-time against the sha256= prefix of X-Impri-Signature.
 *
 * Headers to pass:
 *   - X-Impri-Signature  → signature parameter  (e.g. "sha256=abc...")
 *   - X-Impri-Timestamp  → timestamp parameter   (Unix seconds as string)
 *   - X-Impri-Nonce      → nonce parameter        (hex string)
 *
 * @throws {ImpriWebhookSignatureError} if the signature is missing, malformed,
 *   stale (outside toleranceSec window), or does not match.
 */
export function verifyWebhook(
  rawBody: Buffer | string,
  secret: string,
  timestamp: string,
  nonce: string,
  signature: string,
  opts: { toleranceSec?: number } = {},
): void {
  const toleranceSec = opts.toleranceSec ?? 300

  if (!signature.startsWith('sha256=')) {
    throw new ImpriWebhookSignatureError(
      'X-Impri-Signature header is missing or does not start with "sha256=".',
    )
  }

  const expectedHex = signature.slice(7)

  const ts = parseInt(timestamp, 10)
  if (Number.isNaN(ts)) {
    throw new ImpriWebhookSignatureError(
      'X-Impri-Timestamp header is missing or is not a valid integer.',
    )
  }

  const nowSec = Math.floor(Date.now() / 1000)
  if (Math.abs(nowSec - ts) > toleranceSec) {
    throw new ImpriWebhookSignatureError(
      `Webhook timestamp is outside the ${toleranceSec}s tolerance window ` +
        '(replay protection). Ensure your server clock is synced.',
    )
  }

  const body = Buffer.isBuffer(rawBody) ? rawBody.toString('utf8') : rawBody
  const signingPayload = `${ts}.${nonce}.${body}`
  const actualHex = createHmac('sha256', secret).update(signingPayload, 'utf8').digest('hex')

  // Constant-time comparison to prevent timing side-channel attacks.
  let mismatch = false
  try {
    const expected = Buffer.from(expectedHex, 'hex')
    const actual = Buffer.from(actualHex, 'hex')
    if (expected.length !== actual.length) {
      mismatch = true
    } else {
      mismatch = !timingSafeEqual(expected, actual)
    }
  } catch {
    mismatch = true
  }

  if (mismatch) {
    throw new ImpriWebhookSignatureError(
      'Webhook signature does not match. Verify the secret matches the value from GET /v1/project.',
    )
  }
}

File: integrations/openai-agents/impri_openai/_webhook.py
"""Standalone webhook signature verification.

No client instance required — this is a pure function that uses only stdlib.
"""
from __future__ import annotations

import hashlib
import hmac
import time

from ._errors import ImpriWebhookSignatureError

_EXPECTED_PREFIX = "sha256="


def verify_webhook(
    raw_body: bytes | str,
    secret: str,
    timestamp: str,
    nonce: str,
    signature: str,
    tolerance_sec: int = 300,
) -> None:
    """Verify the HMAC-SHA256 signature of an Impri webhook delivery.

    Algorithm::

        sha256=HMAC-SHA256(secret, f'{timestamp}.{nonce}.{raw_body}')

    Pass the exact raw bytes of the request body (NOT parsed JSON) and the
    three ``X-Impri-*`` headers::

        verify_webhook(
            raw_body=request.body,
            secret=project.webhook_secret,
            timestamp=request.headers['X-Impri-Timestamp'],
            nonce=request.headers['X-Impri-Nonce'],
            signature=request.headers['X-Impri-Signature'],
        )

    Raises:
        ImpriWebhookSignatureError: on HMAC mismatch, missing secret,
            unrecognized signature format, or stale timestamp (replay protection).
    """
    if not secret:
        raise ImpriWebhookSignatureError(
            "webhook_secret must not be empty. "
            "Fetch it with client.get_project() → project['webhook_secret']."
        )
    if not signature.startswith(_EXPECTED_PREFIX):
        raise ImpriWebhookSignatureError(
            f"Unrecognized signature format: {signature!r}. Expected 'sha256=...'."
        )

    # Replay-protection: reject deliveries outside the tolerance window.
    try:
        ts = int(timestamp)
    except (ValueError, TypeError):
        raise ImpriWebhookSignatureError(
            "X-Impri-Timestamp is missing or not a valid integer."
        )
    if abs(time.time() - ts) > tolerance_sec:
        raise ImpriWebhookSignatureError(
            f"Webhook timestamp is outside the {tolerance_sec}s tolerance window "
            "(replay protection). Ensure your server clock is synced."
        )

    if isinstance(raw_body, str):
        raw_body = raw_body.encode()

    payload = f"{timestamp}.{nonce}.".encode() + raw_body
    mac = hmac.new(secret.encode(), payload, hashlib.sha256)
    expected = f"{_EXPECTED_PREFIX}{mac.hexdigest()}"

    if not hmac.compare_digest(expected, signature):
        raise ImpriWebhookSignatureError(
            "Webhook signature mismatch. "
            "Check that you are using the correct webhook_secret and "
            "that the raw request body has not been modified."
        )
